### PR TITLE
Show all template parts using the edited navigation

### PR DESCRIPTION
**Pull request by draganescu**
Co-authored-by: Dennis Snell <5431237+dmsnell@users.noreply.github.com>

## What

This PR shows which template parts use a navigation menu when the menu is edited in isolation in the site editor:

## Screen recording

https://github.com/WordPress/gutenberg/assets/107534/248ad34c-5890-4324-9a53-de637e7ef9ed

### CODE REVIEW

**Comment by draganescu** [comment]
- Should we generalize this so that it works for certain related post types in core: template parts using a navigation, templates using a template part, templates using a pattern, template parts using a pattern?
- The field holds any "related" item, but maybe we should filter by current theme in the response? Currently this is handled by `getEntityRecords`.

This `wp_id` vs `id` in templates and template parts creates a bit of confusion, because it's the only record not addresable by record id. In this case for some reason I could not do an `includes` query using `getEntityRecords`. because the id is the `theme//slug` format not the record id.

**Comment by draganescu** [comment]
I am thinking a better approach here is to extend the search endpoint to support a has_blocks param for post search with optional filtering by attribute name/value. 

This has_blocks param should not augment WP_Query but instead be treated separately as:

- get all posts (of type) limit to a decent default batch
- loop through the batch 
- parse blocks
- find blocks of name (and attributes of value)
- return found posts

This is a very expensive operation, the default limits should be low. On the other hand it's perfectly cache-able. This would allow the generalisation of the feature here to apply to multiple scenarios:

- find all pages that contain the slider block with red background
- find all products which contain a discount block with value set to 10
- find all templates with a query block that has page length 2 🤷🏻

**Comment by getdave** [comment]
_Great_ exploration. Perhaps it would be useful to bring this up in REST API channels and discuss options with them if you haven't already?

**Comment by draganescu** [comment]
I took a week pause from this work due to other work, will resume now.

**Comment by draganescu** [comment]
I've [tested the idea](https://github.com/WordPress/gutenberg/pull/55782/files#diff-00f17538e9365cf3e8d8f46721dcd574112457bb439036a759ecce5ec84d2bb5R12) of a controller that searches all posts based on 

- post type
- block name
- block attribute name
- block attribute value

The main blocker is that it's hard to optimize this sort of search. That controller pages through the DB until it has found enough matching posts. If we want _all_ by default it's not feasible.

I think the other option - also in this PR - with just registering extra rest response fields that expose these relationships between posts based on blocks (template_part using wp_navigation with ref of certain value) is best.

I will continue on this path. I need to see what place is best for registering / adding the extra field.

File: lib/compat/wordpress-6.5/class-block-search-controller.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+				)
+			);
+		}
+
+		/**
+		 * Permission callback to check if the user can use this endpoint.
+		 */
+		public function permissions_check() {
+			// Check user permissions.
+			return true;
```
Comment (dmsnell) [inline lib/compat/wordpress-6.5/class-block-search-controller.php]: this is hopefully a stub. if someone isn't signed in it seems like we wouldn't want them to be able to search the blocks in any way, even if we later restrict the search to posts they can read based on their authentication and capabilities.

File: lib/compat/wordpress-6.5/class-block-search-controller.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			$paged            = 1; // Start on the first page.
+			$posts_with_block = array();
+
+			do {
+				$posts = get_posts(
+					array(
+						'post_type'      => $post_type,
+						'posts_per_page' => $posts_per_page,
+						'paged'          => $paged,
+						'post_status'    => 'publish',
```
Comment (dmsnell) [inline lib/compat/wordpress-6.5/class-block-search-controller.php]: why limit this to published posts? it seems like drafts and scheduled posts would both be important to include.

File: lib/compat/wordpress-6.5/class-block-search-controller.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+						break 2;
+					}
+
+					$blocks       = parse_blocks( $post->post_content );
+					$found_blocks = $this->recursively_find_block_by_attribute(
+						$blocks,
+						$block_name,
+						$attribute_name,
+						$attribute_value
+					);
```
Comment (dmsnell) [inline lib/compat/wordpress-6.5/class-block-search-controller.php]: while we know this is already a potentially huge strain on a server's load, because it has to iterate over all the posts, parsing blocks and recursing and merging arrays only adds to the overhead.

we might consider instead using the block parser's tokenizing regex pattern to scan through posts and note only the opening and void tags, as the nesting and inner HTML and inner blocks aren't important to this search.

```php
function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
	$at = 0;
	$pattern = sprintf(
		'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
		// @TODO: we could ensure that there's a namespace on the block name.
		preg_quote( $block_name, '~' )
	);

	while ( false !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE, $at ) ) {
		$at = $matches[0][1] + strlen( $matches[0][0] );

		if ( ! isset( $attribute_name ) ) {
			return true;
		}

		// Don't parse JSON if it's not possible for the attribute to exist.
		if ( ! str_contains( $matches['attrs'][0], "\"{$attribute_name}\":" ) ) {
			continue;
		}

		$attrs = json_decode( $matches['attrs'][0], /* as-associative */ true );
		if ( ! isset( $attrs[ $attribute_name ] ) ) {
			continue;
		}

		if ( ! isset( $attribute_value ) ) {
			return true;
		}

		if ( $attribute_value === $attrs[ $attribute_name ] ) {
			return true;
		}
	}

	return false;
}
```

this can then be called in an outside loop and it won't carry the memory or performance overhead of actually parsing a post, plus it can stop processing as soon as it finds the requested block instead of traversing the entire post content.

File: lib/compat/wordpress-6.5/class-block-search-controller.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+							),
+							'attribute_name'  => array(
+								'required' => true,
+							),
+							'attribute_value' => array(
+								'required' => true,
+							),
+							'posts_per_page'  => array(
+								'required' => false,
+							),
```
Comment (dmsnell) [inline lib/compat/wordpress-6.5/class-block-search-controller.php]: given how this is used it seems a bit misleading to call this `posts_per_page` when there's no mechanism to grab successive pages and when the behavior is closer to `max_posts` or `max_search_results`

**Review by dmsnell — COMMENTED** [review]
There's quite a timebomb locked inside of this logic. Someone is going to come along and want to add a cache as well. It would be good to provide some guidance on how that could go well or go wrong.

I've suggested a "bandaid" which addresses the inner loop without addressing the broader issue, of needing to scan through a site's posts.

At a minimum it seems like there should be some hard constraints built into this: a time limit imposed when searching, possibly the max number of posts to scan.

If we're relying on this in the editor I could imagine it being a really hard pause when loading for a site with thousands or tens of thousands of posts. We have no guarantee that posts contained the searched-for block exist in the "front" of the query either; it could be 3000 posts and then the first one with our block.

Mitigating this might imply something even more inherent, such as ensuring that the editor expects a progressively loading and asynchronous response. That is, setup an API call and show a spinner and start showing matching posts as they come in, but have the API call poll until it's done. This might involve creating a cursor to associate the search.

**Comment by dmsnell** [comment]
As an idea of an interface that recognizes that this could take tens of seconds or minutes or longer to response, we could create an actual search resource which can be later polled for streaming results.

```
> POST /wp/v2/block-search?block_name=core/cover
< 200 {"results": "/wp/v2/block-search/akdlk35d6sd5"}
```

first matching posts…
```
> GET /wp/v2/block-search/akdlk35d6sd5
< 200 OK {"posts": […], "next-results": "/wp/v2/block-search/akdlk35d6sd5"}
```

still thinking about it, scanning through posts but not finding any matches…
```
> GET /wp/v2/block-search/akdlk35d6sd5
< 200 OK {"posts": [], "next-results": "/wp/v2/block-search/akdlk35d6sd5"}
```

all posts have been scanned and there is nothing left to scan…
```
> GET /wp/v2/block-search/akdlk35d6sd5
< 200 OK {"posts": […]}
```

This gives the client the information it needs to start showing a UI, to grab the next batch of matched posts, to know when to continue polling, and when to stop. There's a problem here in that we don't have guarantees about how long it will take to find, say, the next _five_ posts, so an API response might come at some time-based interval (say, give up and return after 3 seconds) but which should still indicate that the client needs to continue to poll.

Obviously there are concurrency issues with posts being updated in the background, plus the fact that we want to avoid making this query multiple times from the client, but we also need to recognize that with time, the same query might produce different inherent results.

**Comment by draganescu** [comment]
I know fully well the time bomb that is why I am moving slowly here :). Thanks for all the super valuable consideration.

File: lib/compat/wordpress-6.5/class-block-search-controller.php (hunk 1 above)
Comment (draganescu) [reply lib/compat/wordpress-6.5/class-block-search-controller.php]: Yes yes a stub!

File: lib/compat/wordpress-6.5/class-block-search-controller.php (hunk 2 above)
Comment (draganescu) [reply lib/compat/wordpress-6.5/class-block-search-controller.php]: Another stub.

File: lib/compat/wordpress-6.5/class-block-search-controller.php (hunk 4 above)
Comment (draganescu) [reply lib/compat/wordpress-6.5/class-block-search-controller.php]: Good point.

**Comment by draganescu** [comment]
@dmsnell I wonder if the [original approach](https://github.com/WordPress/gutenberg/pull/55782/files/e05357604ef2889eb70e6100b38d5e5f77cc52e7#diff-eff9c92c6df530de32ef7996bdd32fd8de756338cc3ef91e049254658ede14c8) of the PR which aims only to show which template part uses a navigation block, and does that only to scan through template parts, and adds a rest field to the `wp_navigation` resource only is a way forward to fix the problem at hand.

Then I can start a PR for a more generic locator resource that implements the optimisations and documentation you mentioned above.

**Comment by dmsnell** [comment]
> PR which aims only to show which template part uses a navigation block, and does that only to scan through template parts

if it didn't involve creating a new public resource/endpoint/value I would think that's a fair proposal, but there's a risk here in that this only really kicks the can down the road, proverbially. a site with hundreds of template parts will exhibit the same problem and people will say "Gutenberg is not suitable for large agency work."

I recently explored using chunked responses to stream the reply of a long-running API call, but I'm pretty sure that's unworkable with WordPress' "REST" facilities, and it would probably end up timing out at the kinds of delays I would expect this endpoint to return.

creating the generic resource locator may not be as hard as it seems. we might be able to build it with a transient and it only needs to store the offset and query parameters to generate the next query batch/page. that transient can disappear after five minutes and everything will be fine; clients should expect that these cursors might fail or disappear.

**Comment by paaljoachim** [comment]
Thank you for working on this @draganescu Andrei!

It is very helpful to be able to click into a Navigation to get information on where it is being used. 

Next up.
It would be helpful to perhaps have an icon showing header template part beside the Navigation name in the main Navigation list. So one can at quick glance see where the various menus are being used. Then being able to switch the location a menu uses right there in the sidebar without having to go into canvas view.

**Comment by draganescu** [comment]
@paaljoachim that's a good suggestion, but not for this PR, it'd be great to make an issue with a mockup to illustrate your idea which I like quite a lot.

@dmsnell 

> if it didn't involve creating a new public resource/endpoint/value I would think that's a fair proposal, but there's a risk here in that this only really kicks the can down the road, proverbially. a site with hundreds of template parts will exhibit the same problem and people will say "Gutenberg is not suitable for large agency work."

I think register a field, in the site editor or by the navigation block itself, it does not expose anything other way to access this underperformat-by-design feature. I am not too afraid of hundreds of template parts. I could even bail at 50, sorted desc by most recent (b/c you want to see this info in current parts) and filter this number so people who trust their servers can bump it.

I will proceed to remove the public resource and register the rest field in the response for `wp_template_part` as it originally proposed (see [this commit range](https://github.com/WordPress/gutenberg/pull/55782/files/e05357604ef2889eb70e6100b38d5e5f77cc52e7))

**Comment by paaljoachim** [comment]
An issue for brainstorming how to switch menu locations in the Navigation sidebar:
https://github.com/WordPress/gutenberg/issues/56858

**Comment by dmsnell** [comment]
> I could even bail at 50, sorted desc by most recent (b/c you want to see this info in current parts) and filter this number so people who trust their servers can bump it.

This might mean settling for responses with zero results, because we can't guarantee that we will find fifty without scanning through hundreds or thousands of posts before we get the first fifty containing the block.

**Comment by draganescu** [comment]
> This might mean settling for responses with zero results, because we can't guarantee that we will find fifty without scanning through hundreds or thousands of posts before we get the first fifty containing the block.

No, I meant scan only 50. And yes, emtpy results for this particular problem it's fine.

File: lib/compat/wordpress-6.5/class-block-search-controller.php (hunk 3 above)
Comment (dmsnell) [reply lib/compat/wordpress-6.5/class-block-search-controller.php]: @draganescu I've been exploring a lazy parser with significantly better performance characteristics, especially when only wanting to find blocks in a post. it needs work before it can be proposed, but it's great because it does what I proposed here but is memory efficient and can stop once we've found what we're looking for.

WordPress/wordpress-develop#5705

(I do plan on porting this soon into the Gutenberg repository)

File: lib/init.php [hunk 5]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+	 * @return array template parts that use the menu
+	 */
+	function get_template_parts_that_use_menu( $wp_navigation_id ) {
+		// get all wp_template_part posts.
+		$wp_template_part_posts = get_posts(
+			array(
+				'post_type'      => 'wp_template_part',
+				'posts_per_page' => -1,
+			)
+		);
```
Comment (getdave) [inline lib/init.php:139]: I think we'll want to use `WP_Query` here with the various parameters that optimise the query performance such as `no_found_rows`...etc.

File: lib/init.php [hunk 6]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
```
Comment (dmsnell) [inline lib/init.php]: Algorithmically this is no longer recursive, nor is it conceptually. I'm not sure this is helpful to mention "recursive" because it might lead someone astray trying to figure out _how_ it's recursive.

We might say something more to the point about the goal:

> Returns whether the given HTML contains a block of the given type
> and, if provided, a given attribute and attribute value.

File: lib/init.php [hunk 7]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
```
Comment (dmsnell) [inline lib/init.php]: If I were using this I'd be asking a few questions:
 - is the `core/` namespace required or optional?
 - what does it mean to provide an attribute name but not a value?
 - how do I search for a block whose attribute is `null`?

```php
/*
 * …
 *
 * Note that it's not possible to search for an attribute whose value is `null`.
 *
 * @param string  $block_name      Find this block type, with an optional "core/" namespace,
 *                                 e.g. "paragraph", "core/paragraph", "my_plugin/my_block".
 * @param ?string $attribute_name  If provided, the block must also contain this attribute.
 * @param ?mixed  $attribute_value If provided, the given attribute's value must also match this.
 */
```

of course we can find ways of searching for `null`, but I think it's fine not to, and simply to note that it can't be done due to conflating the optional parameter with the definite value `null`

File: lib/init.php [hunk 8]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
```
Comment (dmsnell) [inline lib/init.php:91]: this is a great comment! I would consider adding a link of some form (doesn't need to be a _link_ in fact, but some way to point people to the right place).

```php
/**
 * …
 *
 * @see WP_Block_Parser
 *
 * @var string
 */
```

File: lib/init.php [hunk 9]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			// @TODO: we could ensure that there's a namespace on the block name.
+			preg_quote( $block_name, '~' )
```
Comment (dmsnell) [inline lib/init.php]: this is more or less a decision to make, whether to support the optional core namespace, but it's easy enough to augment.

```php
$full_block_name = str_contains( $block_name, '/' ) ? $block_name : "core/{$block_name}";
```

File: lib/init.php [hunk 10]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+		);
+
+		while ( false !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE, $at ) ) {
+
+			// bail if no matches
```
Comment (dmsnell) [inline lib/init.php]: Not to be the bearer of bad news, but WordPress Coding Standards are going to throw a stink at comments like these. They need to start with a capital letter and end in a `.`

File: lib/init.php [hunk 11]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+			if ( empty ($matches) ) {
+				return false;
+			}
```
Comment (dmsnell) [inline lib/init.php]: did I recommend this code? I may have, but I think `$matches` can never be empty here can it? if that _were_ the case then I believe that `preg_match()` would have returned `false` and we would have exited the loop.

if we're inside the loop it means we matched and we should be guaranteed to have an array of at least one two-element array value (the full match and the full match's offset).

File: lib/init.php [hunk 12]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+			$at = $matches[0][1] + strlen( $matches[0][0] );
+
+			if ( ! isset( $attribute_name ) ) {
+				return true;
+			}
+
+			// Don't parse JSON if it's not possible for the attribute to exist.
+			if ( ! str_contains( $matches['attrs'][0], "\"{$attribute_name}\":" ) ) {
+				continue;
+			}
```
Comment (dmsnell) [inline lib/init.php]: interesting note, but I believe this check can fail if the key is using non-latin1 characters. I wonder if it would actually be better to omit this optimization (which again, sorry for recommending it and then recommending its removal) for now, or note the discrepancy as a clue for someone in the future. we have to avoid situations like `{"\u0061ge": 42}` failing to match for `age`

File: lib/init.php [hunk 13]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+
+			$attrs = json_decode( $matches['attrs'][0], /* as-associative */ true );
+			if ( ! isset( $attrs[ $attribute_name ] ) ) {
```
Comment (dmsnell) [inline lib/init.php:108]: I think this is the place where we'd have to match `null` if we wanted to expose that. `array_key_exists` vs. `isset`. of course, we'd also have to communicate this intent in the function args, so I think this is fine. still, I guess in this moment we're not searching for a value, so if the attribute exists and is set to `null` then we'd be saying that the block with a given attribute name doesn't exist when in fact it does.

File: lib/init.php [hunk 14]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+	 * @param int $wp_navigation_id The menu id.
+	 *
+	 * @return array template parts that use the menu
+	 */
+	function get_template_parts_that_use_menu( $wp_navigation_id ) {
+		// get all wp_template_part posts.
+		$wp_template_part_posts = get_posts(
+			array(
+				'post_type'      => 'wp_template_part',
+				'posts_per_page' => -1,
```
Comment (dmsnell) [inline lib/init.php:137]: it looks like we abandoned all limits here, but this is being added to the query for template parts. there could still be a kind of time-bomb in here because some sites will have hundreds if not thousands of template parts, and we're not able to limit our search. this means that for every request to fetch template parts, we will be processing every template part in the database.

it'd be great if we could impose an arbitrary limit or only add this if a specific query arg requests it so that we don't accidentally make the trivial case obstructively slow, when someone doesn't need to know what template parts use the given menu.

if we limit to an arbitrary amount it would seem like we could add some kind of next-page link or cursor id to continue the search on-demand.

File: lib/init.php (hunk 9 above)
Comment (getdave) [reply lib/init.php]: I'd prefer if we weren't clever about this and just required the block name in full. I don't want to have to look at the code to find out why passing just `navigation` magically works 🙏

File: lib/init.php (hunk 10 above)
Comment (getdave) [inline lib/init.php]:
```suggestion
```

The code is self documenting.

File: lib/init.php (hunk 7 above)
Comment (getdave) [reply lib/init.php]: Could we consider having an array of attributes => values rather than two separate args? In future folks might want to be even more selective about their block but we're going to be baking in to a three argument signature.

The new signature would be:

`(string $block_name, array $attributes)`.

File: lib/init.php (hunk 14 above)
Comment (getdave) [reply lib/init.php:137]: @dmsnell I thought this was only called to get the field value when the REST query is for `wp_navigation` posts? The `get_callback` invokes this function to get the field value for the `template_parts_that_use_menu` field.

How is it being added to the query for template parts? Maybe I missed something?

If I'm correct, then we could still further optimise this approach by only including the field if certain query params are provided. We don't need it for standard requests. I don't know the best REST etiquette for that but some REST maintainers might be able to assist.

File: lib/init.php (hunk 6 above)
Comment (draganescu) [reply lib/init.php]: Ah copy pasta sorry!

File: lib/init.php [hunk 15]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+			)
+		);
+		// loop through them and find the ones that have a navigation block,
+		// with the ref attribute set to $wp_navigation_id.
```
Comment (getdave) [inline lib/init.php]:
```suggestion
		// Loop through template parts and find the ones that have a navigation block,
		// with the ref attribute set to $wp_navigation_id.
```

File: lib/init.php [hunk 16]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+		$wp_template_part_posts_with_navigation = array();
+		foreach ( $wp_template_part_posts as $wp_template_part_post ) {
+			$found_avigation = html_contains_block(
+				$wp_template_part_post->post_content,
+				'navigation',
+				'ref',
+				$wp_navigation_id
+			);
+			if ( $found_avigation ) {
```
Comment (getdave) [inline lib/init.php]:
```suggestion
			$found_navigation = html_contains_block(
				$wp_template_part_post->post_content,
				'navigation',
				'ref',
				$wp_navigation_id
			);
			if ( $found_navigation ) {
```

File: lib/init.php [hunk 17]
```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
...
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
```
Comment (getdave) [inline lib/init.php:82]: I forgot to include in previous review but I think it would be wise to have some tests for this function. You could use a PHPUnit data provider to pass in various combinations of values to ensure you've covered a range of scenarios.

One such example would be what @dmsnell said about Latin characters.

File: lib/init.php (hunk 14 above)
Comment (dmsnell) [reply lib/init.php:137]: > when the REST query is for wp_navigation posts? 

yeah that was my mistake. still, my point is that we're time-bombing the API and I can imagine a handful of normal cases where all we want is the list of `wp_navigation` posts. making the field opt-in would remove that burden and only charge the cost of computation when it's wanted.

this could go _really bad_ on real sites with lots of template parts. say we have a site with a thousand template parts and one navigation menu. even with that single nav menu, we still have to scan all thousand template part posts every time we request the list of nav menus.

File: lib/init.php (hunk 9 above)
Comment (dmsnell) [reply lib/init.php]: this is a fair take, though it can be balanced against the fact that the use of an implicit `core/` namespace permeates Gutenberg, so the choice not to do it here is the surprising behavior. personally I don't care much as long as it's clear in the function documentation, but it's not particularly "clever" to support a well-spec'd and near-universal behavior.

File: lib/init.php (hunk 7 above)
Comment (dmsnell) [reply lib/init.php]: I almost suggested this but on the other hand I feel like this is enough for now and expanding to look for more than one is something that could be addressed in the future, probably with another function.

things in this function are optimized _because_ it's only looking for a single kind of block. if we start allowing multiple search blocks then we would have to start unraveling some of what mitigates the performance issues with this. having a separate function that's costlier and which communicates that is the way out of baking in a simple signature.

wanted to share only because I almost added the same kind of comment you left but didn't, and I wanted to share why I thought it might serve us better as-is with a limited interface.

File: lib/init.php (hunk 11 above)
Comment (draganescu) [reply lib/init.php]: > if we're inside the loop it means we matched 

In this code, if we're inside the loop it means the matcher did not error out.

[preg_match() ](https://www.php.net/manual/en/function.preg-match.php) returns 1 if the pattern matches given subject, 0 if it does not, or false on failure. This code says:

- if the matching did not result in an error
- and it found something
- do this work

In `next_token` in WP_Block_Parse in core the checks are done more explicitly.

EDIT: 

I need to change this anyway b/c the loop is not really happening based on `$at` if we look at the return of `preg_match`. The loop should continue, updating `$at` as long as:

- there are no matches
- the end of the document has not been reached

That means we need to loop based on if `preg_match` returns anything but `0` and treat errors inside.

**Comment by draganescu** [comment]
Excellent, I am happy this is on track now. I will:

- address the feedback about comments and optimisations
- try to move this in private methods (if possile via extending the posts controller)
- add some tests

File: lib/init.php (hunk 11 above)
Comment (dmsnell) [reply lib/init.php]: yeah my bad. I believe we wanted `while ( 1 === preg_match( … ) )` here

Comment (draganescu) [reply lib/init.php]: I did copy paste too much too 😁  and then fixed without considering what I fixed so 🤷🏻

File: lib/init.php (hunk 14 above)
Comment (getdave) [reply lib/init.php:137]: Let's set a hard limit on the number of items returned in the query. However we want to avoid situations where there are a lot of template parts which causes us to miss the key template parts for the site.

To avoid this we could also consider optimising for common template parts such as `header` and `footer` to ensure these are always checked first. It's likely to serve the 80% use case.

**Comment by getdave** [comment]
> ...if possile via extending the posts controller

I have previously received feedback from REST maintainers that extending _existing_ REST endpoints in arbritary ways for specific posts is not considered best practise. This is because (as you know) Gutenberg is not the only consumer of the API.

You may find that a dedicated _block editor specific_ endpoint (such as we implemented for Navigation Fallbacks) would be a better approach. Certainly this is what I was previously advised.

File: lib/init.php (hunk 14 above)
Comment (dmsnell) [reply lib/init.php:137]: > we want to avoid situations where there are a lot of template parts which causes us to miss the key template parts for the site.

doing what we're trying to do here I think is going to require a search index, and one we'll likely need to build and maintain ourselves. I just don't see it being practical to scan the database every time we want to grab `wp_navigation` items.

that can be done, but it's probably easier to do on its own.

> serve the 80% use case

if we _have_ and can _get_ a `header` and `footer` quickly then that would seem like a good idea

**Comment by draganescu** [comment]
Thanks for the useful info @getdave - I'll need to see how that applies to this particular situation. Either way I have no attachment to _how_ more I have it to _what_ which is to hide the html seek function for a while until we figure out the best tweaks to it.

File: lib/init.php (hunk 14 above)
Comment (getdave) [reply lib/init.php:137]: > to scan the database every time we want to grab wp_navigation items.

I agree which is why I think this comes back around to limiting the scope of this change to it's own endpoint for the block editor only. 

This way we do not register this field on the Posts Controller responses _at all_.

We can easily introduce a new private Core Data selector which queries this endpoint only when necessary. 

This will avoid hitting the DB each time there is a query for `wp_navigation` to Posts Controller. 

If we want to apply further safety limits we can do that _as well_.

File: lib/init.php (hunk 7 above)
Comment (draganescu) [reply lib/init.php]: I agree with Dennis on this, from the perspective of keeping this focused is a keystone of our performance mitigating solution, for now. Plus I would really like this method to exist in a non public way, possibly used only by this feature here for now. And this feature does fine with one block one attribute one value.

File: lib/init.php (hunk 9 above)
Comment (draganescu) [reply lib/init.php]: The way this regexp is, we should do more like:

```
preg_quote( str_replace( 'core/', '', $block_name ), '~' );
```

so that both 'navigation' and `core/navigation' work but also `kadence/column`, but `column` means `core/column`, because we omit the core namespace when saving block markup, but it's fine to allow people to be explicit if they want to.

File: lib/init.php (hunk 7 above)
Comment (getdave) [reply lib/init.php]: If it's not public then single attribute is fine for now 👍

File: lib/init.php (hunk 9 above)
Comment (dmsnell) [reply lib/init.php]: yeah @draganescu something happened with the inline code spans in your response, which unfortunately made it hard for me to clearly interpret.

I'll rephrase what I think you might have already denoted:

 - a search for `paragraph` should match on `paragraph` or `core/paragraph`.
 - `<!-- wp:paragraph -->` should match for `paragraph` and `core/paragraph`.
 - a search for `paragraph` _should not_ match on `kadence/paragraph`.
 - `<!-- wp:paragraph -->` _should not_ match for `kadence/paragraph`.

So basically we have to treat all block names without a `core/` namespace as if they had one.

```php
$block_type = str_contains( $block_name, '/' ) ? $block_name : "core/{$block_name}";
preg_match( '... wp:(P<namespace> ... /)?(P<name> ... ) ...', ... );
$found_block_type = empty( $match['namespace'] ) ? "core/{$match['name']}" : $match['name'];

if ( $block_type === $found_block_type ) {
	…
}
```

File: lib/init.php [hunk 18]
```diff
@@ -57,3 +57,121 @@ function gutenberg_menu() {
...
+	 * whose value is `null`.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name      Find this block type,
+	 *                                  with an optional "core/" namespace,
+	 *                                  e.g. "paragraph", "core/paragraph",
+	 *                                  "my_plugin/my_block".
+	 * @param  string  $attribute_name  If provided, the block must also
+	 *                                  contain this attribute.
+	 * @param  string  $attribute_value If provided, the given attribute's
```
Comment (dmsnell) [inline lib/init.php:77]: nitpick: we'll want to add `?string` (I'm pretty sure vs. `string|null`) before merging

File: lib/init.php [hunk 19]
```diff
@@ -57,3 +57,121 @@ function gutenberg_menu() {
...
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 *
+		 * @see WP_Block_Parser.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			preg_quote( str_replace( 'core/', '', $block_name ), '~' )
+		);
```
Comment (dmsnell) [inline lib/init.php:98]: you probably already know this, but we need to expand this a bit because then we'd overlook posts containing `core/template`, for example. although the editor prunes the `core/` namespace, it's perfectly valid to have it written in the HTML.

given the fact that we're only looking for a single block type I recommend we hard-code the regex pattern to match the sought block name. we can do that by breaking apart the block name into its parts, different than the code I shared above about joining them.

```php
// check my string-index math
$solidus_at       = strpos( $block_name, '/' );
$sought_namespace = false === $solidus_at ? 'core' : substr( $block_name, 0, $solidus_at );
$sought_name      = false === $solidus_at ? $block_name : substr( $block_name, $solidus_at + 1 );

$pattern = sprintf(
	'~<!--\s+?wp:((?:%s/)?%s\s+...~',
	preg_quote( $sought_namespace, '~' ),
	preg_quote( $sought_name, '~' )
);
```

**Comment by draganescu** [comment]
I am going to close this for now and come back to it with a better plan or this plan but a fresh implementation.
